### PR TITLE
Feature/00264

### PR DESCRIPTION
**Pull request by IW-SS**
First PR for Energy Domain specific connectors/sources.

Supports the following markets -

- MISO

Added test cases for MISO sources & updated the documentation.

### CODE REVIEW

**Review by ellisk737 — DISMISSED** [review]
changes reviewed & approved

File: src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    query_datetime_format: str = "%Y%m%d"
+    required_options: list = []
+    spark_schema = StructType([StructField("id", IntegerType(), True)])
+
+    def __init__(self, spark: SparkSession, options: dict) -> None:
+        self.spark = spark
+        self.options = options
+        self.current_date = datetime.utcnow()
+
+    def fetch_from_url(self, url_suffix: str) -> bytes:
```
[the remark this answers is not in the document]
Comment (IW-SS) [reply src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py]: Converted into an internal method.

File: src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        response = requests.get(url)
+        code = response.status_code
+
+        if code != 200:
+            raise Exception(f"Unable to access URL `{url}`."
+                            f" Received status code {code} with message {response.content}")
+
+        return response.content
+
+    def pull_data(self) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (IW-SS) [reply src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py]: Converted into an internal method.

File: src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        All the children ISO classes must override this method and call the fetch_url method
+        in it.
+
+        Returns:
+             Raw DataFrame from API.
+        """
+
+        return pd.DataFrame([{"id": 1}])
+
+    def prepare_data(self, df: pd.DataFrame) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (IW-SS) [reply src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py]: Converted into an internal method.

File: src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        Args:
+            df: Raw DataFrame, received from the API.
+
+        Returns:
+             Modified DataFrame, ready for basic use.
+
+        """
+        return df
+
+    def sanitize_data(self, df: pd.DataFrame) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (IW-SS) [reply src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py]: Converted into an internal method.

File: src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        Args:
+            df: Initial modified version of DataFrame, received after preparing the data.
+
+        Returns:
+             Final version of data after all the fixes and modifications.
+
+        """
+        return df
+
+    def get_data(self) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (IW-SS) [reply src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py]: Converted into an internal method.

File: src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        """
+        Spark entrypoint, It executes the entire process of pulling, transforming & fixing data.
+        Returns:
+             Final Spark DataFrame converted from Pandas DataFrame post-execution.
+
+        """
+
+        try:
+            self.pre_read_validation()
+            df = self.spark.createDataFrame(self.get_data(), schema=self.spark_schema)
```
[the remark this answers is not in the document]
Comment (IW-SS) [reply src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/base_iso.py]: Yes expansion is possible.

File: src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/miso_historical_load_iso.py
[the remark this answers is not in the document]
Comment (IW-SS) [reply src/sdk/python/rtdip_sdk/pipelines/sources/spark/iso/miso_historical_load_iso.py:1]: Converted all of those methods into the internal methods. Thanks for the suggestion.

**Review by GBBBAS — APPROVED** [review]

**Pull request merged**